crates/datasources/src/mongodb/mod.rs

```diff
@@ -271,18 +279,136 @@ impl TableProvider for MongoTableProvider {
         // Note that this projection will only project top-level fields. There
         // is not a way to project nested documents (at least when modelling
         // nested docs as a struct).
-        let projected_schema = match projection {
+        let schema = match projection {
             Some(projection) => Arc::new(self.schema.project(projection)?),
             None => self.schema.clone(),
         };
 
-        // TODO: Filters.
+        // Projection document. Project everything that's in the schema.
+        //
+        // The `_id` field is special and needs to be manually suppressed if not
+        // included in the schema.
+        let mut proj_doc = Document::new();
+        let mut has_id_field = false;
+        for field in &schema.fields {
+            proj_doc.insert(field.name(), 1);
+            has_id_field = has_id_field || field.name().as_str() == ID_FIELD_NAME;
+        }
+
+        if !has_id_field {
+            proj_doc.insert(ID_FIELD_NAME, 0);
+        }
+
+        let mut find_opts = FindOptions::default();
+        find_opts.limit = limit.map(|v| v as i64);
+        find_opts.projection = Some(proj_doc);
 
-        Ok(Arc::new(MongoBsonExec::new(
-            projected_schema,
-            self.collection.clone(),
-            limit,
-        )))
+        let filter = exprs_to_mdb_query(_filters).expect("could not build query");
```

Comment:
We'll want to return an error here (e.g. with `map_err(|e| ...)`) to avoid panicking.

Longer term I think we'll want to indicate to the MongoBsonExec whether or not we were successful in creating the filter. If we're not, then we have to the filtering in memory.

Reply:
> Longer term I think we'll want to indicate to the MongoBsonExec whether or not we were successful in creating the filter. If we're not, then we have to the filtering in memory.
I think we'll always want to do the filtering in memory, unless we're successful at creating a filter that _perfectly_ transforms the users' query: the goal of this, I think, is to limit the number of records that data-fusion has to consider, not to obviate data-fusion entirely (for that we'd need/want to write a complete sql-to-mdb query translator, and that's a much much larger undertaking)

Reply:
Actually for this let's do this:

```rust
let filter = match exprs_to_mdb_query(filters).ok()
```

Then down below use `filter` instead of `Some(filter)`. Since we're using `TableProviderFilterPushDown::Inexact`, there's going to be a filter node above this anyways.

Reply:
I think I have something that resembles your intent here; let me know :D